snooty/flutter.py

```diff
@@ -0,0 +1,250 @@
+        super().__init__('Unexpected field: "{}"'.format(bad_field), ty, bad_data)
+        self.bad_field = bad_field
+
+
+def check_type(ty: Type[C], data: object, ty_module: str = '') -> C:
+    # Check for a primitive type
+    if isinstance(ty, type) and issubclass(ty, (str, int, float, bool, type(None))):
+        if not isinstance(data, ty):
+            raise LoadWrongType(ty, data)
+        return cast(C, data)
+
+    # Check for an object
+    if isinstance(data, dict) and ty in CACHED_TYPES:
+        annotations = CACHED_TYPES[ty]
+        if annotations is None:
+            annotations = {field.name: field for field in dataclasses.fields(ty)}
+            CACHED_TYPES[ty] = annotations
+        result: Dict[str, object] = {}
+
+        # Assign missing fields None
+        missing: Set[str] = set()
+        for key in annotations:
+            if key not in data:
+                data[key] = None
+                missing.add(key)
+
+        # Check field types
+        for key, value in data.items():
+            if key not in annotations:
+                raise LoadUnknownField(ty, data, key)
+
+            field = annotations[key]
+            have_value = False
+            if key in missing:
+                # Use the field's default_factory if it's defined
+                try:
+                    result[key] = field.default_factory()  # type: ignore
+                    have_value = True
+                except TypeError:
+                    pass
+
+            if not have_value:
+                result[key] = check_type(field.type, value, ty.__module__)
+
+        output = ty(**result)
+        start_line = getattr(data, '_start_line', None)
+        if start_line is not None:
+            setattr(output, '_start_line', start_line)
+        return output
+
+    # Check for one of the special types defined by PEP-484
+    origin = getattr(ty, '__origin__', None)
+    if origin is not None:
+        args = getattr(ty, '__args__')
+        if origin is list:
+            if not isinstance(data, list):
+                raise LoadWrongType(ty, data)
+            return cast(C, [check_type(args[0], x, ty_module) for x in data])
+        elif origin is dict:
+            if not isinstance(data, dict):
+                raise LoadWrongType(ty, data)
+            key_type, value_type = args
+            return cast(C, {
+                check_type(key_type, k, ty_module): check_type(value_type, v, ty_module)
+                for k, v in data.items()})
+        elif origin is tuple and isinstance(data, collections.abc.Collection):
+            if not len(data) == len(args):
+                raise LoadWrongArity(ty, data)
+            return cast(C, tuple(
+                check_type(tuple_ty, x, ty_module) for x, tuple_ty in zip(data, args)))
+        elif origin is Union:
+            for candidate_ty in args:
+                try:
+                    return cast(C, check_type(candidate_ty, data, ty_module))
```

Comment:
What should happen if more than one candidate in the Union would succeed?  It's sort of the user's fault for creating a Union whose cases overlap, but it could be useful to signal an error in that case instead of picking one case arbitrarily.

For example, in mitx we have some data like this:
```
class MultipleChoice:
    question: str
    choices: List[Choice]

class CheckAllThatApply:
    question: str
    choices: List[Choice]

class Chapter:
    problems: List[Union[MultipleChoice, CheckAllThatApply]]
```

So if we serialized a Chapter and then deserialized it again, it would be helpful to get an error telling us about the ambiguity.  Otherwise it would silently convert all the CheckAllThatApply objects to MultipleChoice.

Reply:
I opted to roughly replicate the behavior of https://serde.rs/enum-representations.html#untagged

The trouble is it's not really possible to detect all ambiguous cases from just the type signature without building a full type space search, and for performance reasons I'd rather not try every possible type candidate since that could be arbitrarily expensive depending on how deep the type definitions are.